File: src/main/java/io/github/nexuskrop/nexusshift/locale/MessageSet.java
package io.github.nexuskrop.nexusshift.locale;

import com.destroystokyo.paper.utils.PaperPluginLogger;
import com.google.common.reflect.TypeToken;
import com.google.gson.GsonBuilder;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.bukkit.Bukkit;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.IOException;
import java.io.InputStreamReader;
import java.util.HashMap;

public final class MessageSet {
    private static final Logger LOGGER = LogManager.getLogger("NS-MessageSet");

    private MessageSet() {
        throw new IllegalStateException("No MessageSet instances for you!");
    }

    private static HashMap<String, String> messages = new HashMap<>();

    public static void load() {
        var gson = new GsonBuilder().enableComplexMapKeySerialization().create();

        var type = new TypeToken<HashMap<String, String>>() {}.getType();

        try (var stream = MessageSet.class.getResourceAsStream("/message.json")) {
            assert stream != null;
            messages = gson.fromJson(new InputStreamReader(stream), type);
        } catch (IOException io) {
            LOGGER.error("Error while reading messages from resources", io);
        }
    }

    public static @Nullable String getOrNull(String key) {
        return messages.get(key);
    }

    public static @NotNull String get(String key) {
        var result = getOrNull(key);
        if (result == null) throw new IllegalArgumentException("No such key");
        return result;
    }
}

File: src/main/kotlin/io/github/nexuskrop/shift/locale/MessageSet.kt
/*
 * Copyright (c) 2022. All rights reserved.
 * Licensed under GNU Affero GPL, either V3 or any later version.
 */

package io.github.nexuskrop.shift.locale

import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.reflect.TypeToken
import io.github.nexuskrop.shift.ui.Messages
import io.github.nexuskrop.shift.ui.NativeComponentSerializer
import io.github.nexuskrop.shift.util.Common
import net.kyori.adventure.text.minimessage.MiniMessage
import net.minecraft.network.chat.Component
import net.minecraft.network.chat.TextComponent
import org.apache.logging.log4j.LogManager
import org.apache.logging.log4j.Logger
import java.io.File
import java.io.InputStream
import java.io.InputStreamReader
import java.nio.charset.StandardCharsets

object MessageSet {
    private val mapGson : Gson = GsonBuilder().enableComplexMapKeySerialization().create()
    private val logger : Logger = LogManager.getLogger("Nexus-MessageSet")
    private var messages : HashMap<String, String> = HashMap<String, String>()

    init {
        throw IllegalStateException("No CommandEngine instances for you!")
    }

    /**
     * Loads the messages from the
     */
    @JvmStatic
    fun loadDefault() {
        loadFromStream(this.javaClass.getResourceAsStream("/messages.json"))
        applyCustomisationFile()
    }

    private fun applyCustomisationFile() {
        val file = File(Common.getDataFolder(), "messages.json")

        if (!file.exists()) {
            logger.info("No customisation file")
            return
        }

        val stream = file.inputStream();

        try {
            val result = mapGson.fromJson<Map<String, String>>(InputStreamReader(stream, StandardCharsets.UTF_8),
                    object : TypeToken<Map<String, String>>(){}.type)
            stream.close()

            if (result.isEmpty()) return

            for ((key,value) in result) {
                messages[key] = value;
            }
        } catch (ex : Exception) {
            logger.warn("Failed to read message customisation file:", ex)
        }
    }

    @JvmStatic
    fun get(key : String) : String {
        return messages[key] ?: key
    }

    @JvmStatic
    fun getMini(key : String) : Component {
        val orig = get(key)

        return if (orig == key) TextComponent(key)
        else {
            NativeComponentSerializer.nativeComponentSerializer().serialize(
                    MiniMessage.miniMessage().deserialize(orig)
            )
        }
    }

    private fun loadFromStream(stream : InputStream?) {
        if (stream == null) {
            logger.warn("Null stream, will use default")
            return
        }

        try {
            val result = mapGson.fromJson<HashMap<String, String>>(InputStreamReader(stream, StandardCharsets.UTF_8),
                    object : TypeToken<HashMap<String, String>>(){}.type)
            stream.close()
            messages = result
        } catch (ex : Exception) {
            logger.error("Failed to read resources:", ex)
        }
    }
}